beacon-chain/sync/rpc_blob_sidecars_by_range.go

```diff
@@ -0,0 +1,186 @@
+				tracing.AnnotateError(span, chunkErr)
+				return wQuota, chunkErr
+			}
+			s.rateLimiter.add(stream, 1)
+			wQuota -= 1
+			// Stop streaming results once the quota of writes for the request is consumed.
+			if wQuota == 0 {
+				return 0, nil
+			}
+		}
+	}
+	return wQuota, nil
+}
+
+// blobsSidecarsByRangeRPCHandler looks up the request blobs from the database from a given start slot index
+func (s *Service) blobSidecarsByRangeRPCHandler(ctx context.Context, msg interface{}, stream libp2pcore.Stream) error {
+	var err error
+	ctx, span := trace.StartSpan(ctx, "sync.BlobsSidecarsByRangeHandler")
+	defer span.End()
+	ctx, cancel := context.WithTimeout(ctx, respTimeout)
+	defer cancel()
+	SetRPCStreamDeadlines(stream)
+	log := log.WithField("handler", p2p.BlobSidecarsByRangeName[1:]) // slice the leading slash off the name var
+
+	r, ok := msg.(*pb.BlobSidecarsByRangeRequest)
+	if !ok {
+		return errors.New("message is not type *pb.BlobsSidecarsByRangeRequest")
+	}
+	if err := s.rateLimiter.validateRequest(stream, 1); err != nil {
+		return err
+	}
+	rp, err := validateBlobsByRange(r, s.cfg.chain.CurrentSlot())
+	if err != nil {
+		s.writeErrorResponseToStream(responseCodeInvalidRequest, err.Error(), stream)
+		s.cfg.p2p.Peers().Scorers().BadResponsesScorer().Increment(stream.Conn().RemotePeer())
+		tracing.AnnotateError(span, err)
+		return err
+	}
+
+	// Ticker to stagger out large requests.
+	ticker := time.NewTicker(time.Second)
+	defer ticker.Stop()
+	batcher, err := newBlockRangeBatcher(rp, s.cfg.beaconDB, s.rateLimiter, s.cfg.chain.IsCanonical, ticker)
+
+	var batch blockBatch
+	wQuota := params.BeaconNetworkConfig().MaxRequestBlobsSidecars
+	for batch, ok = batcher.next(ctx, stream); ok; batch, ok = batcher.next(ctx, stream) {
+		batchStart := time.Now()
+		wQuota, err = s.streamBlobBatch(ctx, batch, wQuota, stream)
+		rpcBlobsByRangeResponseLatency.Observe(float64(time.Since(batchStart).Milliseconds()))
+		if err != nil {
+			return err
+		}
+		// once we have written MAX_REQUEST_BLOB_SIDECARS, we're done serving the request
+		if wQuota == 0 {
+			break
+		}
+	}
+	if err := batch.error(); err != nil {
+		log.WithError(err).Debug("error in BlocksByRange batch")
+		s.writeErrorResponseToStream(responseCodeServerError, p2ptypes.ErrGeneric.Error(), stream)
+		tracing.AnnotateError(span, err)
+		return err
+	}
+
+	closeStream(stream, log)
+	return nil
+}
+
+func blobsByRangeMinStartSlot(current primitives.Slot) (primitives.Slot, error) {
+	minReqEpochs := params.BeaconNetworkConfig().MinEpochsForBlobsSidecarsRequest
+	currEpoch := slots.ToEpoch(current)
+	minStart := params.BeaconConfig().DenebForkEpoch
+	if currEpoch > minReqEpochs && currEpoch-minReqEpochs > minStart {
+		minStart = currEpoch - minReqEpochs
+	}
+	return slots.EpochStart(minStart)
+}
+
+func blobBatchLimit() uint64 {
+	return uint64(flags.Get().BlockBatchLimit / fieldparams.MaxBlobsPerBlock)
```

Comment:
This doesn't seem right. Shouldn't it be the maximum of the two?

Reply:
Ultimately this value controls how many blocks we'll process in a batch, ie this is rate limiting blocks/sec that we query and stream. In the context of `BeaconBlocksByRange`, we stream `batch.size` values (1 blocks for each item in the batch), but for `BlobSidecarsByRange`, we'll stream up to `batch.size * MaxBlobsPerBlock` values (blobs). We want both requests to serve roughly the same number of values per cycle, so we divide the block limit by the number of blobs per block to arrive at roughly the same number of values written to the wire per batch / same interval of time. I should add a comment to better explain what this is for.